### ISSUE TITLE
Two factor authentication required

### ISSUE THREAD

**Issue by HomebridgeX** [merdok/homebridge-miot#119]
Hello, I get this error:
Login error. Two factor authentication required, please use visit below url and retry login.
Thank you!

**Comment by merdok** [comment]
As the error states, you have two factor authentication activated for that reason you need to visit the specified url and allow the connection.

**Comment by HomebridgeX** [comment]
I tried that, but the url changes every time it is called. In addition, I have not found a way to deactivate the 2nd factor. I didn't activate it myself.

**Comment by merdok** [comment]
The url changes because of security reasons, each time you try a new login it will change. You have to either visit the url or disable the 2 factor authentication.

**Comment by HomebridgeX** [comment]
Thanks for the answer. Unfortunately, I can't find a way to deactivate the 2nd factor. I wrote to Mi Support.

**Comment by merdok** [comment]
Yeah, well I never had it activated for my account so I unfortunately cannot help you with that. Please let me know what the mi support will tell you.

**Comment by shahrilamrias** [comment]
> Thanks for the answer. Unfortunately, I can't find a way to deactivate the 2nd factor. I wrote to Mi Support.

Did you get any reply or find any solution? 

For now I just follow the link to auth and then restart HB. No idea how long the login will last until it expires though.

I checked two factor setting in mi dashboard and mine is clearly marked as **disabled**. https://account.xiaomi.com/fe/service/account/home/2step I tried toggling on then off but same result. 

I'm guessing the 2FA we see is not actually 2FA but some sort of security challenge (though I'm guessing they work similarly?). Don't quote me on that though.

**Comment by shahrilamrias** [comment]
Apparently, as it turns out, the two factor setting in mi account dashboard is pretty much useless if we want to disable account verification as they automatically decide whether to show it or not. At least according to the wording in this FAQ https://account.xiaomi.com/helpcenter/faq/en_US/03.account-safety/12.safety-verification/faq-1

I really wish I'm wrong though. Since validating manually via sms/email each time auth expires is gonna be a pain.

**Comment by merdok** [comment]
Yeah, well there is nothing really i cannot do about it... i use the specified credentials to login to the account and in some cases this comes back. I am not sure when or why but there must be some way to disable it? I mean most of the account do not have that active somehow...

**Comment by HomebridgeX** [comment]
Unfortunately I haven't gotten any further with the support. I was told that I should contact my PC for support ???

**Comment by HomebridgeX** [comment]
For me it is also the case that the second factor is definitely deactivated in the account. I never activated it myself. Suddenly it stopped working from one day to the next.

**Comment by merdok** [comment]
That is certainly a mystery when or why this appears. It does not really make sense that for some people it pops up and for some not... Maybe it has something to do with the region?

**Comment by HomebridgeX** [comment]
I'm in Germany.

**Comment by merdok** [comment]
Same here 😅

**Comment by HomebridgeX** [comment]
This is the answer from the support:
"Account verification at login cannot be turned off"

**Comment by merdok** [comment]
Apparently it somehow can be turned off when most of the people do not have it...

**Comment by jojost1** [comment]
I have the same issue after setting up my homebridge instance from scratch. I see the challenge url and I can complete it on the web, but the plugin will never authenticate successfully.. 2FA is 'off' for me as well.

**Comment by merdok** [comment]
It is important to find out why some people have this and most not... I mean there must be a way to somehow disable this? Maybe just new accounts have?

**Comment by shahrilamrias** [comment]
> It is important to find out why some people have this and most not... I mean there must be a way to somehow disable this? Maybe just new accounts have?

Here's some background on my account. 
Maybe it can help narrow the criteria down 
- Never logged on or connected to any Mi mobile devices
- As far as I can tell it was created September 2020
- The verification isn't there on Mi group apps. At least, so far.

**Comment by patowod** [comment]
I am having the same issue. What I found out so far is that the URL delivered in the console (under: Error: Two-factor authentication required, please visit the following url and retry login: …) will only show the option to request a two-factor auth code if I log in in to https://i.mi.com/ or https://account.xiaomi.com before actually loading that url from the console. Then, I get a button to actually send the request and I receive the code via mail. It is accepted, which returns a string "ok" on screen.

Unfortunately, even after restarting homebridge, the error message persists. Under the URL (within console) following lines are printed:

` 
at MiCloud._loginStep2 (/usr/local/lib/node_modules/homebridge-miot/lib/protocol/MiCloud.js:327:13)
at processTicksAndRejections (node:internal/process/task_queues:96:5)
at MiCloud.login (/usr/local/lib/node_modules/homebridge-miot/lib/protocol/MiCloud.js:70:9)
at ZhimiAirPurifierVb2._connectToMiCloud (/usr/local/lib/node_modules/homebridge-miot/lib/protocol/MiotDevice.js:195:9)
[03/01/2022, 22:28:43] Got SIGTERM, shutting down Homebridge...
`

This now seems to loop happily ever after, even though my login credentials work when logging into the above-mentioned accounts and the auth codes are also accepted. The autodetect functionality also fails due to login failure.

Don't know if any of this helps debugging this issue, but it is worth mentioning that this setup worked until recently and nothing has changed on my end since then.

**Comment by merdok** [comment]
The URL which needs to be visited for the two factor authentication is not something that i come up by myself, but this is actually included in the response from the xiaomi server when you try to log in with your account and they request the user to visit it in order to proceed.

Anyway, having a look at the code makes me think that there might be something missing... It was added in a pull request and since i do not have any account with two factor auth, i never could test it by myself...
@patowod could you do me a favor a try to modify the code in MiCloud.js slightly and check if that fixes your issue?
Please locally replace this line of code with: https://github.com/merdok/homebridge-miot/blob/ce67aec040de3bbec4bb861dc3142a80ee2177df/lib/protocol/MiCloud.js#L326
with:
`if (!ssecurity && notificationUrl) {`

I somehow have the feeling that even after visiting the url it will still be included in requests so maybe an additional check is required...

**Comment by patowod** [comment]
@merdok Sorry, I couldn’t answer sooner. Today I finally got around to trying your solution. The good  news is, though, that your idea/solution works! 🎉 I did exactly as you wrote. For anyone trying the same, here is the path where I found the MiCloud.js: 

`/usr/local/lib/node_modules/homebridge-miot/lib/protocol`

After changing the line 326 accordingly, I restarted Homebridge and clicked on the link within the error message. I then initiated sending the 2Factor authentication, which worked flawlessly. Next step: None :) Everything behaved and worked like it used to. Thank you very much @merdok
